### PR TITLE
[CRT] Fix atexit

### PR DESCRIPTION
**Pull request by tkreuzer**
Fixes CORE-17362

### CODE REVIEW

File: sdk/lib/crt/startup/atonexit.c [hunk 1]
```diff
@@ -20,8 +20,8 @@
 
   void __cdecl _lock (int _File);
   void __cdecl _unlock (int _File);
-  
-_PVFV *__onexitbegin;
+
+_PVFV *__onexitbegin = (_PVFV *)(LONG_PTR)-1;
```
Comment (HBelusca) [inline sdk/lib/crt/startup/atonexit.c:24]: Looking at how `__onexitbegin` is used in the code, I noticed two "special" values it also uses: "-1" and "NULL". What is the actual difference in meaning (in the context of `__onexitbegin`˙) of these values?

**Comment by Valeriy-Zainullin** [comment]
I confirm that it fixes the issue.

File: sdk/lib/crt/startup/atonexit.c (hunk 1 above)
Comment (tkreuzer) [reply sdk/lib/crt/startup/atonexit.c:24]: 42

Comment (HBelusca) [reply sdk/lib/crt/startup/atonexit.c:24]: For fucking sake @tkreuzer I'm asking for an actual explanation, not for jokes. It's not the first time you answer such shit in PRs,

**Comment by ThFabba** [comment]
We also (even with this patch applied) seem to be getting these at every process exit:
```
(sdk\lib\rtl\heap.c:2243) HEAP: Trying to free an invalid address FFFFFFFF!
Breakpoint 3 hit
ntdll!RtlFreeHeap+0x1ff:
001b:7c93542f 680d0000c0      push    0C000000Dh
kd> k
ChildEBP RetAddr  
0012ff00 7c5bf79e ntdll!RtlFreeHeap+0x1ff [C:\ROS\reactos\sdk\lib\rtl\heap.c @ 2244]
0012ff14 0040260e msvcrt!free+0x1e [C:\ROS\reactos\sdk\lib\crt\wine\heap.c @ 303]
0012ff28 00401f49 rundll32!__call_atexit+0x5e [C:\ROS\reactos\sdk\lib\crt\startup\atonexit.c @ 45]
0012ffb4 00402161 rundll32!__tmainCRTStartup+0x269 [C:\ROS\reactos\sdk\lib\crt\startup\crtexe.c @ 330]
0012ffc0 7c630aa2 rundll32!wWinMainCRTStartup+0x21 [C:\ROS\reactos\sdk\lib\crt\startup\crtexe.c @ 178]
0012fff0 00000000 kernel32!BaseProcessStartup+0x42 [C:\ROS\reactos\dll\win32\kernel32\client\proc.c @ 463]
```

**Comment by wjk** [comment]
@ThFabba I think that error can be fixed by replacing [this line](https://github.com/reactos/reactos/blob/cd18c4d8fbd3368ae6ed0b95674c9d024fe8a69c/sdk/lib/crt/startup/atonexit.c#L45) with the following:

```c
if (first != ((_PVFV *)(LONG_PTR)-1)
{
    free(first);
}
```

I also think you should introduce a `#define` for that -1 value, for readability’s sake. Hope this helps!

**Comment by zefklop** [comment]
I believe PR #3402 is a better fix :-)

**Comment by binarymaster** [comment]
The problem is in fact has been fixed in master branch by #3402, closing as superseded.

**Pull request closed without merging**